### ISSUE TITLE
Serialization failure in service contract (webapi/AMQP and maybe more)

### ISSUE THREAD

**Issue by thomas-kl1** [AmpersandHQ/magento2-disable-stock-reservation#139]
Error you will face:

"Sources" is not supported. Correct the field name and try again.

It's because if you pass to \Magento\Framework\Webapi\ServiceInputProcessor::convertValue the result of \Magento\Sales\Api\OrderRepositoryInterface::get (or getList), the Order object is badly formatted because of this plugin: \Ampersand\DisableStockReservation\Plugin\Model\OrderRepositoryPlugin. It seems that for some reason, the sourceCode is not serialized when Magento tries to rebuild the Sources extension attribute of the order (Magento\InventorySourceSelectionApi\Api\Data\SourceSelectionItemInterface).

Anyway I don't see any usage of this extension attribute in the code (Sources) so I guess it's safe to remove it and the plugin?

<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Api/etc/extension_attributes.xsd">
    <extension_attributes for="Magento\Sales\Api\Data\OrderInterface">
        <attribute code="sources" type="Magento\InventorySourceSelectionApi\Api\Data\SourceSelectionItemInterface[]" />
    </extension_attributes>
</config>

**Comment by thomas-kl1** [comment]
My fix is to unset the sources: 
$order->getExtensionAttributes()?->setSources([]);

**Comment by tr33m4n** [comment]
Hi @thomas-kl1, do you have replication steps for this issue at all?

Thanks

**Comment by thomas-kl1** [comment]
Yes, pass the result of \Magento\Sales\Api\OrderRepositoryInterface::get to Magento\Framework\Webapi\ServiceInputProcessor::convertValue

**Comment by tr33m4n** [comment]
Hi @thomas-kl1,

I've taken a look at this issue. I can replicate with the following script:
```php
<?php

declare(strict_types=1);

use Magento\Framework\App\Bootstrap;
use Magento\Framework\Reflection\DataObjectProcessor;
use Magento\Framework\Webapi\ServiceInputProcessor;
use Magento\Sales\Api\Data\OrderInterface;
use Magento\Sales\Api\OrderRepositoryInterface;

require __DIR__ . '/app/bootstrap.php';

$bootstrap = Bootstrap::create(BP, $_SERVER);
$objManager = $bootstrap->getObjectManager();

/** @var \Magento\Framework\Webapi\ServiceInputProcessor $serviceInputProcessor */
$serviceInputProcessor = $objManager->get(ServiceInputProcessor::class);
/** @var \Magento\Sales\Api\OrderRepositoryInterface $orderRepository */
$orderRepository = $objManager->get(OrderRepositoryInterface::class);
/** @var \Magento\Framework\Reflection\DataObjectProcessor $dataObjectProcessor */
$dataObjectProcessor = $objManager->get(DataObjectProcessor::class);

$order = $orderRepository->get(20); // Or some other order in the DB
$orderData = $dataObjectProcessor->buildOutputDataArray($order, OrderInterface::class);

$result = $serviceInputProcessor->convertValue($orderData, OrderInterface::class);
```
After further research, it appears that this is a core Magento bug and is reported here https://github.com/magento/magento2/issues/39217. The issue is that using something like `\Magento\Framework\Reflection\DataObjectProcessor::buildOutputDataArray` to "serialize" the objects data will convert all properties to snake case. When `\Magento\Framework\Webapi\ServiceInputProcessor::convertValue` then runs with this data, `\Magento\Framework\Webapi\ServiceInputProcessor::getConstructorData` expects the properties to be in camelcase for constructor hydration. As `\Magento\InventorySourceSelection\Model\Result\SourceSelectionItem` constructor args are not snake case, the hydration will fail, resulting in the errors you encountered above.

As this bug is likely affecting more than just this module, we may just need to hang fire for the official fix. I will look at potentially creating a decorator/proxy that might fix this temporarily in the meantime
